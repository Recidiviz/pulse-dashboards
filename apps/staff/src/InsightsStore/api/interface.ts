// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { ActionStrategy } from "../models/ActionStrategy";
import { ClientEvent } from "../models/ClientEvent";
import { ClientInfo } from "../models/ClientInfo";
import { InsightsConfig } from "../models/InsightsConfig";
import { MetricBenchmark } from "../models/MetricBenchmark";
import {
  ExcludedSupervisionOfficer,
  SupervisionOfficer,
} from "../models/SupervisionOfficer";
import { SupervisionOfficerMetricEvent } from "../models/SupervisionOfficerMetricEvent";
import { SupervisionOfficerSupervisor } from "../models/SupervisionOfficerSupervisor";
import { UserInfo } from "../models/UserInfo";

export type PatchUserInfoProps = { hasSeenOnboarding: boolean };
export interface InsightsAPI {
  init(): Promise<InsightsConfig>;
  actionStrategies(supervisorPseudoId: string): Promise<ActionStrategy>;
  userInfo(userPseudoId: string): Promise<UserInfo>;
  patchUserInfo(
    userPseudoId: string,
    props: PatchUserInfoProps,
  ): Promise<UserInfo>;
  metricBenchmarks(): Promise<Array<MetricBenchmark>>;
  supervisionOfficerSupervisors(): Promise<Array<SupervisionOfficerSupervisor>>;
  officersForSupervisor(
    supervisorPseudoId: string,
  ): Promise<Array<SupervisionOfficer>>;
  excludedOfficersForSupervisor(
    supervisorPseudoId: string,
  ): Promise<Array<ExcludedSupervisionOfficer>>;
  supervisionOfficer(officerPseudoId: string): Promise<SupervisionOfficer>;
  supervisionOfficerMetricEvents(
    officerPseudoId: string,
    metricId: string,
  ): Promise<Array<SupervisionOfficerMetricEvent>>;
  clientInfo(clientPseudoId: string): Promise<ClientInfo>;
  clientEvents(
    clientPseudoId: string,
    endDate: Date,
  ): Promise<Array<ClientEvent>>;
}
