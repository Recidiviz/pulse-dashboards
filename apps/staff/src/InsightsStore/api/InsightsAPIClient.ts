// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import {
  ActionStrategy,
  actionStrategySchema,
  ClientEvent,
  clientEventSchema,
  ClientInfo,
  clientInfoSchema,
  ExcludedSupervisionOfficer,
  excludedSupervisionOfficerSchema,
  insightsConfigSchema,
  MetricBenchmark,
  metricBenchmarkSchema,
  SupervisionOfficer,
  supervisionOfficerFixture,
  SupervisionOfficerMetricEvent,
  supervisionOfficerMetricEventSchema,
  supervisionOfficerSchema,
  SupervisionOfficerSupervisor,
  supervisionOfficerSupervisorSchema,
  supervisionOfficerSupervisorsFixture,
  supervisionOfficerVitalsMetricFixture,
  SupervisionVitalsMetric,
  UserInfo,
  userInfoSchema,
} from "~datatypes";

import { formatDateToISO } from "../../utils";
import type { InsightsStore } from "../InsightsStore";
import {
  ActionStrategySurfacedEvent,
  InsightsAPI,
  PatchUserInfoProps,
} from "./interface";

export class InsightsAPIClient implements InsightsAPI {
  // eslint-disable-next-line no-useless-constructor
  constructor(public readonly insightsStore: InsightsStore) {}

  private get apiStore() {
    return this.insightsStore.rootStore.apiStore;
  }

  private get baseUrl() {
    return `${import.meta.env.VITE_NEW_BACKEND_API_URL}/outliers/${this.tenantId}`;
  }

  get tenantId(): string {
    const { currentTenantId } = this.insightsStore.rootStore;
    if (!currentTenantId) {
      throw new Error(`Attempted to fetch data with undefined tenantId`);
    }
    return currentTenantId;
  }

  async init() {
    const endpoint = `${this.baseUrl}/configuration`;
    const fetchedData = await this.apiStore.get(endpoint);
    return insightsConfigSchema.parse(fetchedData.config);
  }

  async actionStrategies(supervisorPseudoId: string): Promise<ActionStrategy> {
    const endpoint = `${this.baseUrl}/action_strategies/${supervisorPseudoId}`;
    const fetchedData = await this.apiStore.get(endpoint);
    return actionStrategySchema.parse(fetchedData);
  }

  async patchActionStrategies(
    props: ActionStrategySurfacedEvent,
  ): Promise<ActionStrategySurfacedEvent> {
    const endpoint = `${this.baseUrl}/action_strategies/${props.userPseudonymizedId}`;
    const fetchedData = await this.apiStore.patch(endpoint, props);
    const {
      userPseudonymizedId,
      officerPseudonymizedId,
      actionStrategy,
      timestamp,
    } = fetchedData;
    return {
      userPseudonymizedId,
      officerPseudonymizedId,
      actionStrategy,
      timestamp,
    };
  }

  async userInfo(userPseudoId: string): Promise<UserInfo> {
    const endpoint = `${this.baseUrl}/user-info/${userPseudoId}`;
    const fetchedData = await this.apiStore.get(endpoint);
    return userInfoSchema.parse(fetchedData);
  }

  async patchUserInfo(
    userPseudoId: string,
    props: PatchUserInfoProps,
  ): Promise<UserInfo> {
    const endpoint = `${this.baseUrl}/user-info/${userPseudoId}`;
    const fetchedData = await this.apiStore.patch(endpoint, props);
    return userInfoSchema.parse(fetchedData);
  }

  async metricBenchmarks(): Promise<MetricBenchmark[]> {
    const endpoint = `${this.baseUrl}/benchmarks`;
    const fetchedData = await this.apiStore.get(endpoint);
    const benchmarkData = fetchedData.metrics as Array<unknown>;
    return benchmarkData.map((d) => {
      return metricBenchmarkSchema.parse(d);
    });
  }

  async supervisionOfficerSupervisors(): Promise<
    SupervisionOfficerSupervisor[]
  > {
    const endpoint = `${this.baseUrl}/supervisors`;
    const fetchedData = await this.apiStore.get(endpoint);
    const supervisorData = fetchedData.supervisors as Array<unknown>;
    return supervisorData.map((b) =>
      supervisionOfficerSupervisorSchema.parse(b),
    );
  }

  async officersForSupervisor(
    supervisorPseudoId: string,
  ): Promise<Array<SupervisionOfficer>> {
    const endpoint = `${this.baseUrl}/supervisor/${supervisorPseudoId}/officers`;
    const fetchedData = await this.apiStore.get(endpoint);
    const officerData = fetchedData.officers as Array<unknown>;
    return officerData.map((b) => supervisionOfficerSchema.parse(b));
  }

  async excludedOfficersForSupervisor(
    supervisorPseudoId: string,
  ): Promise<Array<ExcludedSupervisionOfficer>> {
    const endpoint = `${this.baseUrl}/supervisor/${supervisorPseudoId}/excluded_officers`;
    const fetchedData = await this.apiStore.get(endpoint);
    const officerData = fetchedData.officers as Array<unknown>;
    return officerData.map((b) => excludedSupervisionOfficerSchema.parse(b));
  }

  async supervisionOfficer(
    officerPseudoId: string,
  ): Promise<SupervisionOfficer> {
    const endpoint = `${this.baseUrl}/officer/${officerPseudoId}`;
    const fetchedData = await this.apiStore.get(endpoint);
    const officerData = fetchedData.officer as unknown;
    return supervisionOfficerSchema.parse(officerData);
  }

  async excludedSupervisionOfficer(
    officerPseudoId: string,
  ): Promise<ExcludedSupervisionOfficer> {
    // TODO: (#6044) Convert "/excluded_officer/" to "/officer/" once all officers are served from the same endpoint.
    const endpoint = `${this.baseUrl}/excluded_officer/${officerPseudoId}`;
    const fetchedData = await this.apiStore.get(endpoint);
    const officerData = fetchedData.officer as unknown;
    return excludedSupervisionOfficerSchema.parse(officerData);
  }

  async supervisionOfficerMetricEvents(
    officerPseudoId: string,
    metricId: string,
  ): Promise<SupervisionOfficerMetricEvent[]> {
    const endpoint = `${this.baseUrl}/officer/${officerPseudoId}/events?metric_id=${metricId}`;
    const fetchedData = await this.apiStore.get(endpoint);
    const eventsData = fetchedData.events as Array<unknown>;
    return eventsData.map((b) => supervisionOfficerMetricEventSchema.parse(b));
  }

  async clientInfo(clientPseudoId: string): Promise<ClientInfo> {
    const endpoint = `${this.baseUrl}/client/${clientPseudoId}`;
    const fetchedData = await this.apiStore.get(endpoint);
    const clientData = fetchedData.client as unknown;
    return clientInfoSchema.parse(clientData);
  }

  async clientEvents(
    clientPseudoId: string,
    endDate: Date,
  ): Promise<Array<ClientEvent>> {
    const endpoint = `${
      this.baseUrl
    }/client/${clientPseudoId}/events?period_end_date=${formatDateToISO(
      endDate,
    )}`;
    const fetchedData = await this.apiStore.get(endpoint);
    const eventsData = fetchedData.events as Array<unknown>;
    return eventsData.map((b) => clientEventSchema.parse(b));
  }

  // TODO #6671 - Instead of using fixture data, make request to endpoint once it is deployed
  async vitalsForSupervisor(
    supervisorPseudoId: string,
  ): Promise<Array<SupervisionVitalsMetric>> {
    const supervisor = supervisionOfficerSupervisorsFixture.find(
      (s) => s.pseudonymizedId === supervisorPseudoId,
    );
    if (!supervisor) return [];
    return supervisionOfficerVitalsMetricFixture.map((metric) => {
      return {
        ...metric,
        vitalsMetrics: metric.vitalsMetrics.filter((metricForOfficer) => {
          const officersForSupervisor = supervisionOfficerFixture
            .filter((o) =>
              o.supervisorExternalIds.includes(supervisor?.externalId),
            )
            .map((o) => o.pseudonymizedId);
          return officersForSupervisor.includes(
            metricForOfficer.officerPseudonymizedId,
          );
        }),
      };
    });
  }

  async vitalsForOfficer(
    officerPseudoId: string,
  ): Promise<Array<SupervisionVitalsMetric>> {
    return supervisionOfficerVitalsMetricFixture.map((metric) => {
      return {
        ...metric,
        vitalsMetrics: metric.vitalsMetrics.filter(
          (metricForOfficer) =>
            officerPseudoId === metricForOfficer.officerPseudonymizedId,
        ),
      };
    });
  }
}
