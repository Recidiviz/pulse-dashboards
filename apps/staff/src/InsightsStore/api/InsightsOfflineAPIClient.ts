/* eslint-disable class-methods-use-this */
// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { differenceInDays, subDays } from "date-fns";

import type { InsightsStore } from "../InsightsStore";
import { ClientEvent } from "../models/ClientEvent";
import { ClientInfo } from "../models/ClientInfo";
import { MetricBenchmark } from "../models/MetricBenchmark";
import { LATEST_END_DATE } from "../models/offlineFixtures/constants";
import { leadershipUserInfoFixture } from "../models/offlineFixtures/UserInfoFixture";
import { SupervisionOfficer } from "../models/SupervisionOfficer";
import { SupervisionOfficerMetricEvent } from "../models/SupervisionOfficerMetricEvent";
import { SupervisionOfficerSupervisor } from "../models/SupervisionOfficerSupervisor";
import { UserInfo } from "../models/UserInfo";
import { InsightsAPI, PatchUserInfoProps } from "./interface";

export class InsightsOfflineAPIClient implements InsightsAPI {
  private pseudoIdToEditableUserInfo: Map<
    string,
    { metadata: PatchUserInfoProps }
  > = new Map();

  // eslint-disable-next-line no-useless-constructor
  constructor(public readonly insightsStore: InsightsStore) {}

  async init() {
    const config = (
      await import("../models/offlineFixtures/InsightsConfigFixture")
    ).InsightsConfigFixture;

    return config;
  }

  async userInfo(userPseudoId: string): Promise<UserInfo> {
    const { supervisionOfficerSupervisorsFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerSupervisor"
    );
    const matchingSupervisor = supervisionOfficerSupervisorsFixture.find(
      (supervisor) => supervisor.pseudonymizedId === userPseudoId,
    );

    if (matchingSupervisor) {
      return {
        role: "supervision_officer_supervisor",
        entity: matchingSupervisor,
        ...(this.pseudoIdToEditableUserInfo.get(userPseudoId) ?? {
          metadata: {
            hasSeenOnboarding: false,
          },
        }),
      };
    }

    return {
      ...leadershipUserInfoFixture,
      ...(this.pseudoIdToEditableUserInfo.get(userPseudoId) ?? {
        metadata: {
          hasSeenOnboarding: false,
        },
      }),
    };
  }

  async patchUserInfo(
    userPseudoId: string,
    props: PatchUserInfoProps,
  ): Promise<UserInfo> {
    this.pseudoIdToEditableUserInfo.set(userPseudoId, { metadata: props });
    return this.userInfo(userPseudoId);
  }

  async metricBenchmarks(): Promise<MetricBenchmark[]> {
    const { metricBenchmarksFixture } = await import(
      "../models/offlineFixtures/MetricBenchmarkFixture"
    );
    return metricBenchmarksFixture;
  }

  async supervisionOfficerSupervisors(): Promise<
    SupervisionOfficerSupervisor[]
  > {
    const { supervisionOfficerSupervisorsFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerSupervisor"
    );
    return supervisionOfficerSupervisorsFixture;
  }

  async officersForSupervisor(
    supervisorPseudoId: string,
  ): Promise<Array<SupervisionOfficer>> {
    const { supervisionOfficerFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerFixture"
    );

    return supervisionOfficerFixture.filter(
      (o) => `hashed-${o.supervisorExternalId}` === supervisorPseudoId,
    );
  }

  async supervisionOfficer(
    officerPseudoId: string,
  ): Promise<SupervisionOfficer> {
    const { supervisionOfficerFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerFixture"
    );

    const officerFixture = supervisionOfficerFixture.find(
      (o) => o.pseudonymizedId === officerPseudoId,
    );

    if (!officerFixture)
      throw new Error(`Officer ${officerPseudoId} not present in fixture data`);

    return officerFixture;
  }

  async supervisionOfficerMetricEvents(
    officerPseudoId: string,
    metricId: string,
  ): Promise<SupervisionOfficerMetricEvent[]> {
    const { supervisionOfficerMetricEventFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerMetricEventFixture"
    );
    return supervisionOfficerMetricEventFixture.filter(
      (e) => e.metricId === metricId,
    );
  }

  async clientInfo(clientPseudoId: string): Promise<ClientInfo> {
    const { clientInfoFixture } = await import(
      "../models/offlineFixtures/ClientInfoFixture"
    );

    const clientInfo = clientInfoFixture[clientPseudoId];

    if (!clientInfo) {
      throw new Error(`Client ${clientPseudoId} not present in fixture data`);
    }

    return clientInfo;
  }

  async clientEvents(
    clientPseudoId: string,
    endDate: Date,
  ): Promise<Array<ClientEvent>> {
    const { clientEventFixture } = await import(
      "../models/offlineFixtures/ClientEventFixture"
    );

    // adjust fixture dates relative to the input
    const dateOffsetInDays = differenceInDays(LATEST_END_DATE, endDate);

    return clientEventFixture.map((event) => ({
      ...event,
      eventDate: subDays(event.eventDate, dateOffsetInDays),
    }));
  }
}
