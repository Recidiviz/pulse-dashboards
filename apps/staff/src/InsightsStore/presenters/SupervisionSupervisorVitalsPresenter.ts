// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { ascending } from "d3-array";
import { computed, flowResult, makeAutoObservable } from "mobx";

import {
  ExcludedSupervisionOfficer,
  SupervisionOfficer,
  VitalsMetricForOfficer,
} from "~datatypes";
import { Hydratable, HydratesFromSource } from "~hydration-utils";

import { InsightsSupervisionStore } from "../stores/InsightsSupervisionStore";
import {
  SupervisorVitalsMetricDetail,
  VitalsMetricDetailForOfficer,
} from "./types";
import { labelForVitalsMetricId } from "./utils";

/**
 * The `SupervisionSupervisorVitalsPresenter` class is responsible for managing and presenting
 * data related to a supervisor's officers' vitals metrics within the Recidiviz platform.
 * It handles data hydration, data aggregation, and user-specific contextual information.
 */
export class SupervisionSupervisorVitalsPresenter implements Hydratable {
  // ==============================
  // Properties and Constructor
  // ==============================

  private hydrator: HydratesFromSource;

  constructor(
    protected supervisionStore: InsightsSupervisionStore,
    public supervisorPseudoId: string,
  ) {
    makeAutoObservable(
      this,
      { vitalsMetricDetails: computed },
      { autoBind: true },
    );

    this.hydrator = new HydratesFromSource({
      expectPopulated: this.expectPopulated(),
      populate: async () => {
        await Promise.all(this.populateMethods());
      },
    });
  }

  // ==============================
  // Component specific logic
  // ==============================

  mergeOfficerNames(
    metrics: VitalsMetricForOfficer[],
  ): VitalsMetricDetailForOfficer[] {
    const metricDetails: VitalsMetricDetailForOfficer[] = [];
    metrics.forEach((metric) => {
      const displayName = this.allOfficers.find(
        (o) => o.pseudonymizedId === metric.officerPseudonymizedId,
      )?.displayName;
      if (displayName)
        metricDetails.push({
          ...metric,
          displayName,
        });
    });
    return metricDetails;
  }

  get vitalsMetricDetails(): SupervisorVitalsMetricDetail[] {
    const metrics = this.supervisionStore.vitalsMetricsByPseudoId.get(
      this.supervisorPseudoId,
    );
    if (!metrics) return [];

    return metrics.map((metric) => {
      const sortedMetrics = [...metric.vitalsMetrics].sort((a, b) =>
        ascending(a.metricValue, b.metricValue),
      );
      return {
        label: labelForVitalsMetricId(metric.metricId),
        officersWithMetricValues: this.mergeOfficerNames(sortedMetrics),
      };
    });
  }

  // TODO(#6453) - you don't need any of the bottom two in this presenter.
  // TODO #6617 - the next three methods along with their expectPopulated methods will move
  // to SupervisionSupervisorPagePresenter during the refactor so the redundancy will be removed.
  /**
   * Combines and returns all officers, both included and excluded, under this supervisor.
   * @returns An array of `SupervisionOfficer` and `ExcludedSupervisionOfficer`, or `undefined`.
   */
  get allOfficers(): (SupervisionOfficer | ExcludedSupervisionOfficer)[] {
    return [
      ...(this.officersWithOutcomesData || []),
      ...(this.excludedOfficers || []),
    ];
  }

  /**
   * Provides a list of all officers in this supervisor's unit that were not
   * explicitly excluded from outcomes.
   * @returns An array of `SupervisionOfficer` or `undefined` if data is not available.
   */
  get officersWithOutcomesData(): SupervisionOfficer[] | undefined {
    return this.supervisionStore.officersBySupervisorPseudoId
      .get(this.supervisorPseudoId)
      ?.filter((o) => o.includeInOutcomes === true);
  }

  // TODO(#6453): this should look at and filter officersBySupervisorPseudoId
  /**
   * Provides a list of all officers excluded from outcomes in this supervisor's unit.
   * @returns An array of `ExcludedSupervisionOfficer` or `undefined` if data is not available.
   */
  get excludedOfficers(): ExcludedSupervisionOfficer[] | undefined {
    return this.supervisionStore.excludedOfficersBySupervisorPseudoId.get(
      this.supervisorPseudoId,
    );
  }

  // ==============================
  // Hydration and Initialization
  // ==============================

  /**
   * Returns an array of promises representing the methods required to populate
   * the necessary data for this presenter.
   */
  populateMethods() {
    return [
      flowResult(
        this.supervisionStore.populateVitalsForSupervisor(
          this.supervisorPseudoId,
        ),
      ),
      flowResult(
        this.supervisionStore.populateOfficersForSupervisor(
          this.supervisorPseudoId,
        ),
      ),
      flowResult(
        this.supervisionStore.populateExcludedOfficersForSupervisor(
          this.supervisorPseudoId,
        ),
      ),
    ];
  }

  /**
   * Returns an array of expectations for whether the necessary data has been populated.
   */
  expectPopulated() {
    return [
      this.expectVitalsForSupervisorPopulated,
      this.expectOfficersWithOutcomesPopulated,
      this.expectExcludedOfficersPopulated,
    ];
  }

  private expectVitalsForSupervisorPopulated() {
    if (
      !this.supervisionStore.vitalsMetricsByPseudoId.get(
        this.supervisorPseudoId,
      )
    )
      throw new Error(
        `Failed to populate vitals metrics for supervisor ${this.supervisorPseudoId}`,
      );
  }

  /**
   * Asserts that officers with outcomes have been populated.
   * @throws An error if officers with outcomes are not populated.
   */
  private expectOfficersWithOutcomesPopulated() {
    if (
      !this.supervisionStore.officersBySupervisorPseudoId.has(
        this.supervisorPseudoId,
      )
    )
      throw new Error("failed to populate officers with outcomes");
  }

  /**
   * Asserts that excluded officers have been populated.
   * @throws An error if excluded officers are not populated.
   */
  private expectExcludedOfficersPopulated() {
    if (
      !this.supervisionStore.excludedOfficersBySupervisorPseudoId.has(
        this.supervisorPseudoId,
      )
    )
      throw new Error("failed to populate excluded officers");
  }

  /**
   * Initiates hydration for all data needed within this presenter class
   */
  async hydrate(): Promise<void> {
    return this.hydrator.hydrate();
  }

  get hydrationState() {
    return this.hydrator.hydrationState;
  }
}
