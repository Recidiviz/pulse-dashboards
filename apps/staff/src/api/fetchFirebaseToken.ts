// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

export async function fetchFirebaseToken(
  auth0Token: string,
  impersonationParams?: {
    impersonatedEmail: string;
    impersonatedStateCode: string;
  },
): Promise<string> {
  const queryParams = impersonationParams
    ? `?${new URLSearchParams({
        impersonationParams: JSON.stringify(impersonationParams),
      })}`
    : "";
  const url = `${import.meta.env.VITE_API_URL}/token${queryParams}`;

  const response = await fetch(url, {
    headers: {
      Authorization: `Bearer ${auth0Token}`,
    },
  });

  if (!response.ok) {
    throw new Error(
      `Fetching ${impersonationParams ? "impersonated " : ""}Firebase token failed.\nStatus: ${response.status} - ${response.statusText}`,
    );
  }
  const { firebaseToken } = await response.json();
  return firebaseToken;
}
