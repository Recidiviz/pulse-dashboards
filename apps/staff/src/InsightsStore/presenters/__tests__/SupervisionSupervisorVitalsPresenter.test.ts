// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { flowResult } from "mobx";

import {
  InsightsConfigFixture,
  supervisionOfficerFixture,
  supervisionOfficerSupervisorsFixture,
} from "~datatypes";

import { RootStore } from "../../../RootStore";
import UserStore from "../../../RootStore/UserStore";
import { InsightsOfflineAPIClient } from "../../api/InsightsOfflineAPIClient";
import { InsightsSupervisionStore } from "../../stores/InsightsSupervisionStore";
import { SupervisionSupervisorVitalsPresenter } from "../SupervisionSupervisorVitalsPresenter";

let store: InsightsSupervisionStore;
const pseudoId = "hashed-agonzalez123";

beforeEach(() => {
  vi.spyOn(UserStore.prototype, "userPseudoId", "get").mockImplementation(
    () => pseudoId,
  );
  vi.spyOn(UserStore.prototype, "isRecidivizUser", "get").mockImplementation(
    () => false,
  );

  store = new InsightsSupervisionStore(
    new RootStore().insightsStore,
    InsightsConfigFixture,
  );
  const testSupervisorPseudoId =
    supervisionOfficerSupervisorsFixture[0].pseudonymizedId;

  presenter = new SupervisionSupervisorVitalsPresenter(
    store,
    testSupervisorPseudoId,
  );
});

afterEach(() => {
  vi.restoreAllMocks();
});

let presenter: SupervisionSupervisorVitalsPresenter;

describe("with vitals data already hydrated", () => {
  beforeEach(async () => {
    await Promise.all([
      flowResult(
        store.populateVitalsForSupervisor(presenter.supervisorPseudoId),
      ),
      flowResult(
        store.populateOfficersForSupervisor(presenter.supervisorPseudoId),
      ),
    ]);
  });

  test("is immediately hydrated", () => {
    expect(presenter.hydrationState.status).toBe("hydrated");
  });

  test("makes no additional API calls", async () => {
    vi.spyOn(InsightsOfflineAPIClient.prototype, "vitalsForSupervisor");

    await presenter.hydrate();

    expect(
      store.insightsStore.apiClient.vitalsForSupervisor,
    ).not.toHaveBeenCalled();
  });

  test("has vitalsMetricDetails", () => {
    expect(presenter.vitalsMetricDetails).toBeDefined();
    expect(presenter.vitalsMetricDetails).toMatchSnapshot();
  });

  test("vitalsMetricDetails when there are no VitalsMetricForOfficer for a metric", () => {
    vi.spyOn(InsightsOfflineAPIClient.prototype, "vitalsForSupervisor");
    store.vitalsMetricsBySupervisorPseudoId.set(pseudoId, [
      {
        metricId: "timely_contact",
        vitalsMetrics: [
          {
            officerPseudonymizedId:
              supervisionOfficerFixture[0].pseudonymizedId,
            metric30DDelta: 1,
            metricValue: 99,
          },
        ],
      },
    ]);
    expect(presenter.vitalsMetricDetails).toMatchInlineSnapshot(`
      [
        {
          "officersWithMetricValues": [
            {
              "displayName": "Walter Harris",
              "metric30DDelta": 1,
              "metricValue": 99,
              "officerPseudonymizedId": "hashed-so1",
            },
          ],
          "titleDisplayName": "Timely Contact",
        },
      ]
    `);
  });

  test("vitalsMetricDetails when officer not found", () => {
    vi.spyOn(InsightsOfflineAPIClient.prototype, "vitalsForSupervisor");
    store.vitalsMetricsBySupervisorPseudoId.set(pseudoId, [
      {
        metricId: "timely_contact",
        vitalsMetrics: [
          {
            officerPseudonymizedId: "not-a-real-id",
            metric30DDelta: 1,
            metricValue: 99,
          },
        ],
      },
    ]);
    expect(presenter.vitalsMetricDetails).toMatchInlineSnapshot(`
      [
        {
          "officersWithMetricValues": [],
          "titleDisplayName": "Timely Contact",
        },
      ]
    `);
  });
});

test("throw an when vitalsMetricDetails does not find a vitalsMetricConfig for a metricId", async () => {
  vi.spyOn(InsightsOfflineAPIClient.prototype, "vitalsForOfficer");

  store.vitalsMetricsBySupervisorPseudoId.set(pseudoId, [
    {
      // @ts-ignore
      metricId: "fake_id",
      vitalsMetrics: [
        {
          officerPseudonymizedId: supervisionOfficerFixture[0].pseudonymizedId,
          metric30DDelta: 1,
          metricValue: 99,
        },
      ],
    },
  ]);

  expect(
    () => presenter.vitalsMetricDetails,
  ).toThrowErrorMatchingInlineSnapshot(
    `[Error: Missing configuration for metric fake_id]`,
  );
});

test("hydration", async () => {
  vi.spyOn(InsightsOfflineAPIClient.prototype, "vitalsForSupervisor");

  expect(presenter.hydrationState.status).toBe("needs hydration");

  await presenter.hydrate();

  expect(presenter.hydrationState.status).toBe("hydrated");
  expect(store.insightsStore.apiClient.vitalsForSupervisor).toHaveBeenCalled();
});
