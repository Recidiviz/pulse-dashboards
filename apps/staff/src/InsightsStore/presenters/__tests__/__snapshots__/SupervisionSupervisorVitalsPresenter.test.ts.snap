// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`with vitals data already hydrated > has vitalsMetricDetails 1`] = `
[
  {
    "label": "F2F Contact",
    "officersWithMetricValues": [
      {
        "displayName": "Jason Nelson",
        "metric30DDelta": 1.2,
        "metricValue": 31,
        "officerPseudonymizedId": "hashed-so3",
      },
      {
        "displayName": "Jack Hernandez",
        "metric30DDelta": -1,
        "metricValue": 57,
        "officerPseudonymizedId": "hashed-so2",
      },
      {
        "displayName": "Harriet Boyd",
        "metric30DDelta": 5.9,
        "metricValue": 79,
        "officerPseudonymizedId": "hashed-so9",
      },
      {
        "displayName": "Walter Harris",
        "metric30DDelta": -7,
        "metricValue": 87,
        "officerPseudonymizedId": "hashed-so1",
      },
      {
        "displayName": "Stephen Mann",
        "metric30DDelta": 1.5,
        "metricValue": 89,
        "officerPseudonymizedId": "hashed-so10",
      },
    ],
  },
  {
    "label": "Timely Risk Assessment",
    "officersWithMetricValues": [
      {
        "displayName": "Stephen Mann",
        "metric30DDelta": -3.4,
        "metricValue": 79,
        "officerPseudonymizedId": "hashed-so10",
      },
      {
        "displayName": "Jack Hernandez",
        "metric30DDelta": -1.7,
        "metricValue": 86,
        "officerPseudonymizedId": "hashed-so2",
      },
      {
        "displayName": "Jason Nelson",
        "metric30DDelta": 0.7,
        "metricValue": 97,
        "officerPseudonymizedId": "hashed-so3",
      },
      {
        "displayName": "Walter Harris",
        "metric30DDelta": -4,
        "metricValue": 99,
        "officerPseudonymizedId": "hashed-so1",
      },
      {
        "displayName": "Harriet Boyd",
        "metric30DDelta": 0,
        "metricValue": 100,
        "officerPseudonymizedId": "hashed-so9",
      },
    ],
  },
]
`;
